docs/DafnyRef/Statements.md

```diff
@@ -1739,8 +1746,92 @@ RevealStmt =
+assertion is assumed in the context of the rest of the program.  But with the label, the first assertion is hidden from
+the rest of the program. That assertion can be _revealed_ by adding a `reveal` statement:
+
+```dafny
+method m(i: int) {
+  assert x: i == 0; // Fails
+  reveal x;
+  assert i == 0; // Now succeeds
+}
+```
+At the point of the `reveal` statement, the labeled assertion is made visible and can be used in proving the second assertion.
+In this example there is no point to labeling an assertion and then immediately revealing it. More useful are the cases where
+the reveal is in an assert-by block or much later in the method body.
+
+### 20.20.2. Revealing preconditions
+
+In the same way as assertions, preconditions can be labeled.
+Within the body of a method, a precondition is an assumption; if the precondition is labeled then that assumption is not visible in the body of the method.
+A `reveal` statement naming the label of the precondition then makes the assumption visible.
+
+Here is a toy example:
+```
+method m(x: int, y: int) returns (z: int)
+  requires L: 0 < y;
+  ensures z == x+y
+  ensures x < z
+{
+  z := x + y;
+}
+```
+The above methhod will not verify. In particular, the second postcondition cannot be proved.
+However, if we add a `reveal L;` statement in the body of the method, then the precondition is visible 
+and both postconditions can be proved.
+
+One could also use this style:
+```
+method m(x: int, y: int) returns (z: int)
+  requires L: 0 < y;
+  ensures z == x+y
+  ensures x < z
+{
+  z := x + y;
+  assert x < z by { reveal L; }
+}
+```
+
+The reason to possibly hide a precondition is the same as the reason to hide assertions: 
+sometimes less information is better for the solver as it helps the solver focus attention on 
+relevant information.
+
+Section 7 of [http://leino.science/papers/krml276.html](http://leino.science/papers/krml276.html) provides 
+an extended illustration of this technique.
+
+### 20.20.3. Revealing function bodies
+
+Normally function bodies are transparent and available for constructing proofs of assertions that use those functions.
+However, sometimes it is helpful to mark a function _opaque_ and treat it as an uninterpreted function, whose properties are
+just its specifications.  This action limits the information available to the logical reasoning engine and may make a proof 
+possible where there might be information overload otherwise.
+
+But then there may be specific instances where the definition of that opaque function is needed. In that situation, the
+body of the function can be _revealed_ using the reveal statement. Here is an example:
+```dafny
+function {: opaque} f(i: int): int { i + 1 }
+
+method m(int i) {
+  assert f(i) == i + 1;
+}
+```
+Without the `{:opaque}` attribute, the assertion is valid; with the attribute it cannot be proved because the body if the
```

Comment:
```suggestion
Without the [`{:opaque}`](#sec-opaque) attribute, the assertion is valid; with the attribute it cannot be proved because the body if the
```